feat: show tree input in forms and filters for 'ontology tree' like data.

mswertz opened the pull request:
Motivation: we want to show CollectionEvents.AreasOfInformation as tree input in forms, so cohort data managers can more easily select the appropriate element as compared to current list,  and in data explorer, so researchers can more easily set filters.

Therefore added 'ontology tree'. Also as side effect, merged 'columnFormat' and 'columnType' into one parameter, because having two will lead to errors.

How to test
- in styleguide view, take a look at OntologyTerm component. Notice it is quite responsive.
- in schema designer you can now enter a columnFormat to Collectionevent.areasOfInformation column
- in molgenis.csv of cohort catalogue set columnFormat=OntologyTerm for column=AreasOfInformation to do this during upload
- in forms you can now select areasOfInformation in a tree
- in data explorer you can select filter for areasOfInformation in a tree

Known limitations: this component should be limited to ontologies that are not too large. E.g. for ICD10 we should think to make it more 'lazy'.

![image](https://user-images.githubusercontent.com/120060/130358142-5ec5f0fa-399d-4ce1-8ebd-137e5a3aa1ab.png)

connoratrug commented on apps/styleguide/src/forms/InputOntology.vue:
```diff
@@ -0,0 +1,174 @@
+ * - adding search to pre-filter elements / paths
+ */
+export default {
+  extends: _baseInput,
```
Add name

mswertz replied:
Is that always required? I tend to only add 'name' when the component can be used in a nested way.

connoratrug commented on apps/styleguide/src/forms/InputOntology.vue:
```diff
@@ -0,0 +1,174 @@
+<template>
+  <FormGroup v-bind="$props" v-on="$listeners">
+    <div class="p-0 m-0">
+      <InputOntologySubtree
+        class="pl-0"
+        :terms="terms"
+        :selection="selection"
+        :expanded="expanded"
+        @toggleExpand="toggleExpand"
+        @select="select"
+        @deselect="deselect"
+      />
+    </div>
+  </FormGroup>
+</template>
```
combine with existing Tree component , code duplication

mswertz replied:
Agreed. Did you assess if they can be drop in replacement of each other?

mswertz replied:
> Agreed. Did you assess if they can be drop in replacement of each other?

If I look I get impression that difference is that this new component also takes care of the loading of data. This can be a nice benefit but should check if we also need a 'dumb' component that you can use without preloading and where you inject the data as a prop.

mswertz replied:
merged

connoratrug reviewed (changes requested):
![Screenshot 2021-08-23 at 13 35 06](https://user-images.githubusercontent.com/47183404/130440756-1b5e5c81-05f3-4d80-b6ac-a91860a14891.png)

style guide does not show anything

mswertz commented:
I am kinda warming up to the idea to make this a 'smart' component.

I.e. in case of 'ref' and 'ref_array' AND refTable has parent column THEN show TreeInput

However, that might be too much magic.

mswertz commented:
> ![Screenshot 2021-08-23 at 13 35 06](https://user-images.githubusercontent.com/47183404/130440756-1b5e5c81-05f3-4d80-b6ac-a91860a14891.png)
> 
> style guide does not show anything

Ah, that is because you need a backend to be functional. So the feature request is here: give an error message if backend is not available with suggestion to sign in. Meanwhile you can sign-in in a backend and update the code example if necessary. (styleguide assumed you first ran 'gradle test' and then signed in.

connoratrug commented:
> Ah, that is because you need a backend to be functional. So the feature request is here: give an error message if backend is not available with suggestion to sign in. Meanwhile you can sign-in in a backend and update the code example if necessary. (styleguide assumed you first ran 'gradle test' and then signed in.

The style guide should 'just work' for it to be of any real use. Suggest not to mix the fetching of the data with the ui component, that way the data can be added to the example(s). If for some reason the component needs to fetch the data itself ( generally a bad idea ( reuse , testing, tight coupling ) , then suggest to mock the response, and/or include instruction on how to wire the component ( although this kind of defeats the purpose of having a component demo site).

connoratrug commented:
![Screenshot 2021-08-23 at 14 37 12](https://user-images.githubusercontent.com/47183404/130448248-e81f4924-9c26-40c1-93f7-aacb45643613.png)
After login component still a bit non functional, but maybe the example should be updated

mswertz commented:
> > Ah, that is because you need a backend to be functional. So the feature request is here: give an error message if backend is not available with suggestion to sign in. Meanwhile you can sign-in in a backend and update the code example if necessary. (styleguide assumed you first ran 'gradle test' and then signed in.
> 
> The style guide should 'just work' for it to be of any real use. Suggest not to mix the fetching of the data with the ui component, that way the data can be added to the example(s). If for some reason the component needs to fetch the data itself ( generally a bad idea ( reuse , testing, tight coupling ) , then suggest to mock the response, and/or include instruction on how to wire the component ( although this kind of defeats the purpose of having a component demo site).

I think we have different view of what the scope and purpose is of this component library. In my mind we are not aiming for a general purpose component library. Instead, the whole point of our component library is to work effectively with our backend, so the backend integration is a key element (mocking data is then silly). So I don't want to mock responses (I find mocking a waste of code giving false sense of quality and only reason to mock is when your test become too slow). 

Meanwhile I agree with your 'just works' argument. Generally we therefore include a demo dataset in MOLGENIS (pet store) for this purpose. So we could expand this testset with appropriate example and make that public. And I do agree it should give proper error messages and instruction what to do to fix it (in this case: to sign in). 

And certainly, there are cases where you want to use possibly same component with, and without, the default backend logic. I am not sure this applies to this component though: do we concretely have the need to run it with different query than it now provides out of the box? Otherwise only reason to mock the data is because users don't know how to sign in? Is that not a waste of code (I could add it if you insist, I don't have strong feelings about adding the option to use the component without backend, I only have strong feelings that at least we should also add demo that includes the backend).

mswertz commented:
> ![Screenshot 2021-08-23 at 14 37 12](https://user-images.githubusercontent.com/47183404/130448248-e81f4924-9c26-40c1-93f7-aacb45643613.png)
> After login component still a bit non functional, but maybe the example should be updated

Correct, the example data is indeed "limited". Will update to a more exciting set.

mswertz commented:
I have update pull request
* can now also pass the options via :options 
* also added the components to be used in variable explorer (I must say I really don't get the vuex stuff still but I figured it out) => might now not need keywords load anymore
* added example data to pet store so it works out of box
* added and improve error messages in case you still have to sign in

mswertz commented:
![](https://github.trello.services/images/mini-trello-icon.png) [feat: OntologyInput](https://trello.com/c/tDZ9k96y/215-feat-ontologyinput)

connoratrug reviewed (changes requested):
Looks nice, few small issue; 

Two types of overflow issues ( screenshots below) and

- clicking the leaf label does not toggle the item
- missing on hover action ( show affordance) 

<img width="495" alt="Screenshot 2021-08-25 at 13 56 54" src="https://user-images.githubusercontent.com/47183404/130786639-7efafb4d-e4b5-4d2d-80ed-97090bef06d3.png">
and
<img width="639" alt="Screenshot 2021-08-25 at 13 56 42" src="https://user-images.githubusercontent.com/47183404/130786658-8ad75408-3517-4d35-a2fa-cefd2d54e370.png">

mswertz commented:
Changes:
* fixed overflow by setting different columns width depending screen size
* merge columnType and columnFormat
* add cursor = hand using role=button so cursor is changed mouse over ontology input
* click ontology node toggles expand, click ontology leaf toggles select
